[mlir] Add FileRange location type.

jpienaar opened the pull request:
This location type represents a contiguous range inside a file. It is
effectively a pair of FileLineCols. Add new type and make FileLineCol a
view for case where it matches existing previous one.

The location includes filename and optional start line & col, and end
line & col. Considered common cases are file:line, file:line:col,
file:line:start_col to file:line:end_col and general range within same
file. In memory its encoded as trailing objects. This keeps the memory
requirement the same as FileLineColLoc today (makes the rather common
File:Line cheaper) at the expense of extra work at decoding time. I made
the line/columns accessors uniformly return optionals. Kept the unsigned
type.

There was the option to always have file range be castable to
FileLineColLoc. This cast would just drop other fields. That may result in
some simpler staging. TBD.

This is a rather minimal change, it does not yet add bindings (C or Python),
lowering to LLVM debug locations etc. that supports end line:cols.

joker-eph commented on mlir/test/IR/locations.mlir:
```diff
@@ -16,7 +16,7 @@ func.func @inline_notation() -> i32 {
   // CHECK: } loc(fused["foo", "mysource.cc":10:8])
   affine.for %i0 loc("IVlocation") = 0 to 8 {
-  } loc(fused["foo", "mysource.cc":10:8])
+  } loc(fused["foo", range("mysource.cc":10:8 to 5)])
```
You add an input, but it's lost in the output (no CHECK)?

Technius commented on mlir/unittests/IR/LocationTest.cpp:
```diff
@@ -0,0 +1,54 @@
+
+  llvm::MemoryBufferRef resBuffer(os.str(), "result");
+  llvm::line_iterator it(resBuffer);
+  size_t ltIndex = -1, carrotIndex = -2;
```
nit: `carrot` (the vegetable) is probably supposed to be `caret` (the `^` character).

Technius commented on mlir/include/mlir/IR/BuiltinLocationAttributes.td:
````diff
@@ -102,6 +102,44 @@ def FileLineColLoc : Builtin_LocationAttr<"FileLineColLoc"> {
+
+    ```
+    filelinecol-location ::= `range(` string-literal `:` integer-literal `:`
+                             integer-literal ` to ` integer-literal `)`
````
If the 3rd integer here is a size/length, `to` is misleading in the syntax.

Technius commented on mlir/include/mlir/IR/BuiltinLocationAttributes.td:
```diff
@@ -102,6 +102,44 @@ def FileLineColLoc : Builtin_LocationAttr<"FileLineColLoc"> {
+    inside a file.
+  }];
+  let parameters = (ins "StringAttr":$filename, "unsigned":$line,
+                        "unsigned":$column, "unsigned":$byte_size);
```
What's the motivation for using line/column for the starting position and then using byte size for the end?

In my understanding, tools that report user-facing diagnostics would want to report line/column information. This would require the tool to query the `SourceMgr` (if one is available) for the line/column of the end position, which could result in incorrect column numbers if there are multibyte characters and no information about the file's text encoding is known. Accurate column number calculations would require the original source files to be available to the tool, which may not always be the case. Wouldn't it be better to also store line/column for the end position, instead computing the length dynamically?

If having 4 integers is a concern, how about using two integers (byte offset and length) instead? This would move the concern of text encodings back to the user-facing tool, which will need to translate the byte offset to starting position and (byte offset + length) to ending position.

jpienaar replied:
Yes that's in the PR description

> In particular storing the end line and end column in addition or instead of size. But storing both results in encoding redundant info and wasn't sure the visual benefit in the non-sourgemanager case is worth it. And storing these instead would result in less efficient error emission

For user facting tools, you would be using the SourceMgr or something else (probably a language level equivalent of SourceMgr like clang does) indeed and being able to query these. This output is for the plain command line testing tools/compiler devs.

Now the option of just using offsets for both and start is there or, as you mention, start offset + size is there. It would differ from FileLocation reporting & difficult for users to find start. So the reason I am proposing hybrid, is that it would be as useful as FileLocation for basic tools users while enabling the reporting as range for user facing tools without any redudant encoding. 

 > This would require the tool to query the SourceMgr (if one is available) for the line/column of the end position, which could result in incorrect column numbers if there are multibyte characters and no information about the file's text encoding is known.

The file encoding doesn't change does it? So if the size is computed from the file, then the reporting using the same assumption as during coding should produce same range.

jpienaar commented:
River and I were just talking and he mentioned to also consider debug info generation, which does push more towards the multiple line:col approach. So let me try doing that so we can diff.

jpienaar committed:
[mlir] Add FileRange location type.

This location type represents a contiguous range inside a file. It is
effectively a pair of FileLineCols. Add new type and make FileLineCol a
view for case where it matches existing previous one.

The location includes filename and optional start line & col, and end
line & col. Considered common cases are file:line, file:line:col,
file:line:start_col to file:line:end_col and general range within same
file.

jpienaar commented:
Updated - had a bit too much fun considering different encodings.

jpienaar committed:
Update mlir/include/mlir/IR/Location.h

Co-authored-by: River Riddle <riddleriver@gmail.com>

jpienaar committed:
Update mlir/include/mlir/IR/BuiltinLocationAttributes.td

Co-authored-by: River Riddle <riddleriver@gmail.com>

jpienaar commented on mlir/include/mlir/IR/BuiltinLocationAttributes.td:
```diff
@@ -60,46 +60,97 @@ def CallSiteLoc : Builtin_LocationAttr<"CallSiteLoc"> {
   ];
+
+  let extraClassDeclaration = [{
+    std::optional<unsigned> getStartLine() const;
+    std::optional<unsigned> getStartColumn() const;
+    std::optional<unsigned> getEndColumn() const;
+    std::optional<unsigned> getEndLine() const;
```
SG, I considered this but thought it may be simpler to be explicit. But don't have good rationale for when it may be needed, so switching to simpler. I don't want to expose the in memory encoding, so doing some extra work in the bytecode writer.

jpienaar replied:
Mmm, on second thought. I'm not sure if not using optional doesn't result in requiring more folks to be aware of the convention or encoding. Now if I have question "is this actually a range", a consumer would have to know the above. Which might not be an issue if all consumers treat empty ranges as locations.

River707 replied:
I was trying to map this to all of the other source ranges I know of (SM, clang, lsp, etc.), none of which use optional. In which situation are you thinking this might cause an issue? I haven't yet really encountered anything problematic so far when using the other source ranges I know of.

jpienaar replied:
Its just difference between single location vs range that needs to be differentiated really (and I could just use an isa on FileLineColLoc for that I realized this morning). For the hard coding between bytecode format and what's in memory, we could consider that fine as internal.

jpienaar committed:
Addressing comments

jpienaar committed:
Forgot to fix formatting

jpienaar committed:
Update printer and parser

jpienaar committed:
Fix erroneous conditional

jpienaar committed:
Keep textual output format matching old
